src/standalone/Vegetation/canopy_parameterizations.jl

```diff
@@ -23,31 +23,197 @@ export plant_absorbed_ppfd,
+
+    p₁ = b + μ̄ * h
+    p₂ = b - μ̄ * h
+    p₃ = b + μ̄ * K
+    p₄ = b - μ̄ * K
+
+    d₁ = p₁ * (u₁ - μ̄ * h) / s₁ - p₂ * (u₁ + μ̄ * h) * s₁
+    d₂ = (u₂ + μ̄ * h) / s₁ - (u₂ - μ̄ * h) * s₁
+
+    # h coefficients for direct upward flux
+    h₁ = -d * p₄ - c * f
+    h₂ =
+        1 / d₁ * (
+            (d - h₁ / σ * p₃) * (u₁ - μ̄ * h) / s₁ -
+            p₂ * s₂ * (d - c - h₁ / σ * (u₁ + μ̄ * K))
+        )
+    h₃ =
+        -1 / d₁ * (
+            (d - h₁ / σ * p₃) * (u₁ + μ̄ * h) * s₁ -
+            p₁ * s₂ * (d - c - h₁ / σ * (u₁ + μ̄ * K))
+        )
+
+    # h coefficients for direct downward flux 
+    h₄ = -f * p₃ - c * d
+    h₅ =
+        -1 / d₂ *
+        (h₄ * (u₂ + μ̄ * h) / (σ * s₁) + (u₃ - h₄ / σ * (u₂ - μ̄ * K)) * s₂)
+    h₆ =
+        1 / d₂ *
+        (h₄ / σ * (u₂ - μ̄ * h) * s₁ + (u₃ - h₄ / σ * (u₂ - μ̄ * K)) * s₂)
+
+    # h coefficients for diffuse upward flux
+    h₇ = c * (u₁ - μ̄ * h) / (d₁ * s₁)
+    h₈ = -c * s₁ * (u₁ + μ̄ * h) / d₁
+
+    # h coefficients for diffuse downward flux
+    h₉ = (u₂ + μ̄ * h) / (d₂ * s₁)
+    h₁₀ = -s₁ * (u₂ - μ̄ * h) / d₂
+
+    # Compute the LAI per layer for this canopy
+    Lₗ = LAI / n_layers
+
+    # Initialize FAPAR value and layer counter
+    FAPAR = 0
+    i = 0
+
+    # Intialize vars to save computed fluxes from each layer for the next layer
+    I_dir_up_prev = 0
+    I_dir_dn_prev = 0
+    I_dif_up_prev = 0
+    I_dif_dn_prev = 0
+
+    # Compute FAPAR in each canopy layer
+    while i <= n_layers
+
+        # Compute cumulative LAI at this layer
+        L = i * Lₗ
+
+        # Compute the direct fluxes into/out of the layer 
+        I_dir_up =
+            h₁ * exp(-K * L * Ω) / σ +
+            h₂ * exp(-h * L * Ω) +
+            h₃ * exp(h * L * Ω)
+        I_dir_dn =
+            h₄ * exp(-K * L * Ω) / σ +
+            h₅ * exp(-h * L * Ω) +
+            h₆ * exp(h * L * Ω)
+
+        # Add collimated radiation to downard flux
+        I_dir_dn += exp(-K * L * Ω)
+
+        # Compute the diffuse fluxes into/out of the layer
+        I_dif_up = h₇ * exp(-h * L * Ω) + h₈ * exp(h * L * Ω)
+        I_dif_dn = h₉ * exp(-h * L * Ω) + h₁₀ * exp(h * L * Ω)
+
+        # Energy balance giving radiation absorbed in the layer 
+        if i == 0
+            I_dir_abs = 0
+            I_dif_abs = 0
+        else
+            I_dir_abs = I_dir_up - I_dir_up_prev - I_dir_dn + I_dir_dn_prev
+            I_dif_abs = I_dif_up - I_dif_up_prev - I_dif_dn + I_dif_dn_prev
+        end
+
+        # Add PAR absorbed in the layer to total APAR
+        FAPAR += (1 - diff_perc) * I_dir_abs + (diff_perc) * I_dif_abs
+
+        # Save input/output values to compute energy balance of next layer 
+        I_dir_up_prev = I_dir_up
+        I_dir_dn_prev = I_dir_dn
+        I_dif_up_prev = I_dif_up
+        I_dif_dn_prev = I_dif_dn
+
+        # Move on to the next layer
+        i += 1
+    end
+
+    # Convert FAPAR into APAR and return
+    return PAR * FAPAR
```

Comment:
you might double check that this returns Float64 if the input args are float64 & likewise for float32.